Scaling wiki manual (#2424)
DOMjudge/domjudge: doc/manual/install-domserver.rst

means that access logs for the webserver will still report the IP of the proxy
or loadbalancer.

Scaling the domserver
---------------------

Documenting all possible ways to properly size the domserver is out of scope for this
manual. But properly think about the memory management for the PHP-FPM children with

> ~40 per gigabyte of memory, 500 for a 16Gb system.

Reviewer: This is not clearly phrased. Does it mean to say that 40 children consume 1 GB of memory? Is this a suggestion for how many to configure?

Author: Its copied from the php pool snippet, and it should say something about `total amount of memory in gigabytes divided by 40`, I fully agree its not clearly phrased but I can't think of a better fix which also reads natural.

> * Consider the amount of submissions you expect at any given time, does the load stay uniform.

Reviewer: ```suggestion
* Consider the amount of submissions you expect at any given time, does the load stay uniform?
```

> * Should you expect all teams to solve the easy problems in the first hour.

Reviewer: ```suggestion
* Do you expect all teams to solve the easy problems in the first hour?
```

>   as a rule of thumb you can expect: ``#testcases * #submissions/teams`` as extra load. Similar

Reviewer: ```suggestion
  As a rule of thumb you can expect: ``#testcases * #submissions/teams`` as extra load. Similar
```
Not really clear what "extra load" means here.

Author: What is intended is to address the POST per testcase in a very short amount of time.

means that access logs for the webserver will still report the IP of the proxy
or loadbalancer.

Scaling the domserver
---------------------

Documenting all possible ways to properly size the domserver is out of scope for this
manual. But properly think about the memory management for the PHP-FPM children, we advice

> 40 FPM children per gigabyte of RAM memory so around 500 for a 16Gb system.

Reviewer 2: I wonder if these numbers are still accurate? I think when I came up with this number I just ran htop on a running server (maybe make it busy by forcing some judgehosts to do some rejudging) and noted the rss for the fpm processes, and rounded to some nice number. But maybe it'd be good to double check that.

Author: I think I can do that after the finals, sounds like a reasonable way for testing.

Reviewer 3: Though I think 40/GiB is fine (we use a similar metric at $work, though the workload there is quite different) it might be better to change this to number of requests per second. The section title (which is great!) is regarding scaling domserver, but the body of the text is about how to fully utilize your/a server by running domserver. The wiki-link to 'scaling and load testing' does a better job, though even that link is inconsistent with the statement here (taking 70MiB as a 'more modern' number).

To me it makes more sense to base the requirements of the domserver on the (maximum) number of concurrent users and/or requests, because the number of fpm worker is the limiting factor of how many requests domserver can handle concurrently. Deriving a baseline for this is also much easier since we can simply calculate the number of concurrent (php) requests from the access log. (timestamp + request duration)

Having said that. I think this is fine for now. Lets use the finals to gather some data.

nit: isn't it `we advise` instead of `we advice`?

* Consider the amount of submissions you expect at any given time, does the load stay uniform?
* Do you expect all teams to solve the easy problems in the first hour?

>   As a rule of thumb you can expect: ``#testcases * #submissions/teams`` as extra load (HTTP traffic
>   from the judgehost reporting the results per testcase additionally on the team refreshing the page
>   to know the verdict). Similar for the harder problems (which look simple) where multiple wrong

Reviewer 2: This is really hard to understand. I'm also not sure why dividing by the number of teams?
Maybe something like:
```suggestion
    Do not forget to account for "hidden load", i.e. requests that the judging system makes to the domserver. You can expect ``#testcases * #submissions/teams`` **requests/sec (or what unit is this?)** that you need to account for. Additionally, the scoreboard pages for spectators and teams auto-refresh every 30s (**not sure if this is still true, but I suspect it is**), so every 30 teams = 1 req/sec minimum (on top of teams refreshing manually to know the verdict from their submission).
```

I'm not sure how much to write here, but lazy evaluation is a big factor too.

Author: the `/`was not meant for dividing but more a `or` at this point the number of submissions is equal to the number of teams (1 per team for the simple one)

I fully agree that your way of reading it makes more sense as it mixes up 2 uses for the `/` symbol.

Reviewer 3: I agree with @nickygerritsen, dropping `/teams` makes it more legible here. To iterate on ubergeek's suggestion

```suggestion
Do not forget to account for "hidden load", i.e. requests that the judging system makes to the domserver. You can expect ``#testcases * #submissions`` requests that you need to account for. Not only are these requests spread out over the duration of the contest but they can also be spread out unevenly. The 'middle three' hours of a full five hour contest usually see fewer submissions than the first and last hours. 

This "hidden load" can be mitigated by keeping the "lazy evaluation" option enabled. Lazy evaluation stops whenever the judgehost decides a verdict can be reached. e.g. a submission who's verdict on the first sample testcase is "Wrong Answer" does not need to be checked against the other testcases. Lazy evaluation helps when it is expected that many submissions fail which is often the case.

Additionally, the scoreboard pages for spectators and teams auto-refresh every 30s, so every 30 teams = 1 req/sec minimum (on top of teams refreshing manually to know the verdict from their submission).
```
